bip-0360.mediawiki

```diff
@@ -0,0 +1,630 @@
+cryptosystems.
+
+In the distant future, following the implementation of the P2QRH output type in a QuBit soft fork, there will likely
+be a need for Pay to Quantum Secure (P2QS) addresses. A distinction is made between cryptography that's merely resistant
+to quantum attack, and cryptography that's secured by specialized quantum hardware. P2QRH is resistant to quantum
+attack, while P2QS is quantum secure. These will require specialized quantum hardware for signing, while still
+[https://quantum-journal.org/papers/q-2023-01-19-901/ using public keys that are verifiable via classical means].
+Additional follow-on BIPs will be needed to implement P2QS. However, until specialized quantum cryptography
+hardware is widespread, quantum resistant addresses should be an adequate intermediate solution.
+
+== Specification ==
+
+We define the signature scheme and transaction structure as follows.
+
+=== Descriptor Format ===
+
+To integrate P2QRH into existing wallet software and scripts, we introduce a new output descriptor function
+<code>qrh()</code>. This function represents a P2QRH output, similar to how <code>wpkh()</code> and <code>tr()</code>
+are used for P2WPKH and P2TR outputs, respectively.
+
+The <code>qrh()</code> function takes the HASH256 of the concatenated HASH256 of the quantum-resistant public keys as
+its argument. For example:
+
+  qrh(HASH256(HASH256(pubkey1) <nowiki>||</nowiki> HASH256(pubkey2) <nowiki>||</nowiki> ...))
+
+This function allows wallets to manage P2QRH addresses and outputs while accommodating multiple public keys of varying
+lengths, such as in multisig schemes, while keeping the public keys hidden until the time of spending. At a minimum,
+there should be two public keys in a P2QRH output, one that makes use of classical cryptography and one that makes use
+of a PQC algorithm chosen within the wallet.
+
+=== Address Format ===
+
+P2QRH uses SegWit version 3 outputs, resulting in addresses that start with <code>bc1r</code>, following
+[https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32 BIP-173]. Bech32 encoding maps version 3 to the
+prefix <code>r</code>.
+
+Example P2QRH address:
+
+<code>bc1r...</code> (32-byte Bech32m-encoded HASH256 of the HASH256 of the public keys)
+
+=== ScriptPubKey ===
+
+The <code>scriptPubKey</code> for a P2QRH output is:
+
+  OP_PUSHNUM_3 OP_PUSHBYTES_32 <nowiki><hash></nowiki>
+
+Where:
+
+* <code>OP_PUSHNUM_3</code> (<code>0x03</code>) indicates SegWit version 3.
+* <nowiki><hash></nowiki> is the 32-byte HASH256 of the merkle root of a tree of public key hashes, as defined in the
+Hash Computation section.
+
+=== Output Mechanics ===
+
+To address the risk of arbitrary data being stored using P2QRH (QuBit) outputs, very specific rules will be applied
+to spending from the witness stack in SegWit v3 outputs. A fixed signature size will be necessary for spending the
```

Comment:
This is confusing. We don’t "spend from the witness stack". From what I understand meanwhile, the public keys and public key hashes are provided in the attestation, and any signatures are provided in the witness. Under the assumption that the threshold of necessary keys is pre-committed and I understand that right, how about:

```suggestion
To prevent storage of arbitrary data using P2QRH (QuBit) outputs,
the witness stack for inputs spending segwit v3 outputs is limited to the fixed-size signatures necessary for spending the
```